Makefile

```diff
@@ -62,64 +64,69 @@ BUILD_IMAGE ?= golang:1.9-alpine
+	    -w /go/src/$(PKG)                                                         \
+	    $(BUILD_IMAGE)                                                            \
+	    /bin/sh -c "                                                              \
+	        ARCH=$(ARCH)                                                          \
+	        OS=$(OS)                                                              \
+	        VERSION=$(VERSION)                                                    \
+	        PKG=$(PKG)                                                            \
+	        ./build/build.sh                                                      \
 	    "
 
 # Example: make shell CMD="-c 'date > datefile'"
 shell: build-dirs
 	@echo "launching a shell in the containerized build environment"
-	@docker run                                                             \
-	    -ti                                                                 \
-	    --rm                                                                \
-	    -u $$(id -u):$$(id -g)                                              \
-	    -v "$$(pwd)/.go:/go"                                                \
-	    -v "$$(pwd):/go/src/$(PKG)"                                         \
-	    -v "$$(pwd)/bin/$(ARCH):/go/bin"                                    \
-	    -v "$$(pwd)/bin/$(ARCH):/go/bin/$$(go env GOOS)_$(ARCH)"            \
-	    -v "$$(pwd)/.go/std/$(ARCH):/usr/local/go/pkg/linux_$(ARCH)_static" \
-	    -w /go/src/$(PKG)                                                   \
-	    $(BUILD_IMAGE)                                                      \
+	@docker run                                                                   \
+	    -ti                                                                       \
+	    --rm                                                                      \
+	    -u $$(id -u):$$(id -g)                                                    \
+	    -v "$$(pwd)/.go:/go"                                                      \
+	    -v "$$(pwd):/go/src/$(PKG)"                                               \
+	    -v "$$(pwd)/bin/$(OS)-$(ARCH):/go/bin"                                    \
+	    -v "$$(pwd)/bin/$(OS)-$(ARCH):/go/bin/$(OS)_$(ARCH)"                      \
+	    -v "$$(pwd)/.go/std/$(OS)-$(ARCH):/usr/local/go/pkg/$(OS)_$(ARCH)_static" \
+	    -w /go/src/$(PKG)                                                         \
+	    $(BUILD_IMAGE)                                                            \
 	    /bin/sh $(CMD)
 
 DOTFILE_IMAGE = $(subst :,_,$(subst /,_,$(IMAGE))-$(VERSION))
 
 container: .container-$(DOTFILE_IMAGE) container-name
-.container-$(DOTFILE_IMAGE): bin/$(ARCH)/$(BIN) Dockerfile.in
+.container-$(DOTFILE_IMAGE): bin/$(OS)-$(ARCH)/$(BIN) Dockerfile.in
 	@sed \
 	    -e 's|ARG_BIN|$(BIN)|g' \
+	    -e 's|ARG_OS|$(OS)|g' \
 	    -e 's|ARG_ARCH|$(ARCH)|g' \
 	    -e 's|ARG_FROM|$(BASEIMAGE)|g' \
 	    Dockerfile.in > .dockerfile-$(ARCH)
```

Comment:
I think you also need an `$(OS)` identifier here and the next line to ensure OS/Arch pairs are correctly copied.

Reply:
Not sure if it's necessary as we assume all Docker images are Linux, and as far as I can tell it only affects the name of the .dockerfile-... files. 

Might be better for consistency's sake, but the rest of the container naming convention only includes ARCH, and changing that would conflict with names of images pushed before.

Reply:
That's fair, I was conflating my desire to use this to build darwin (and eventually windows) binaries, but those would not would not be well supported in the `.container-*` targets

Reply:
know little about running Docker on non-linux, but darwin, windows, or any supported GOOS/GOARCH should build at the moment.